### ISSUE TITLE
NullReferenceException on an early EOS

### ISSUE THREAD

**Issue by mauricio-morales** [b3b00/csly#223]
I found this situation where the lexer appears to recognize an EOS before the stream is actually done.
This is happening on a Lucene-like query parser we are putting together. 

This query runs fine:

    (java AND dotnet)

But this one fails

    (java AND dotnet )
                    ^

Somehow, that space triggers an EOS before reading the `)`. I've seen the EOS while debugging, but the error is an NRE:

```
ParserTests.EarlyEos.EarlyEosTests.Early_EOS_NRE

System.NullReferenceException: Object reference not set to an instance of an object.

System.NullReferenceException
Object reference not set to an instance of an object.
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 91
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 89
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 89
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 89
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 89
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(SyntaxNode`1 node, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 89
   at sly.parser.generator.visitor.EBNFSyntaxTreeVisitor`2.Visit(ISyntaxNode`1 n, Object context) in csly/sly/parser/generator/visitor/EBNFSyntaxTreeVisitor.cs:line 30
   at sly.parser.generator.visitor.SyntaxTreeVisitor`2.VisitSyntaxTree(ISyntaxNode`1 root, Object context) in csly/sly/parser/generator/visitor/SyntaxTreeVisitor.cs:line 147
   at sly.parser.Parser`2.ParseWithContext(IList`1 tokens, Object parsingContext, String startingNonTerminal) in csly/sly/parser/parser/Parser.cs:line 103
   at sly.parser.Parser`2.ParseWithContext(String source, Object context, String startingNonTerminal) in csly/sly/parser/parser/Parser.cs:line 85
   at sly.parser.Parser`2.Parse(String source, String startingNonTerminal) in csly/sly/parser/parser/Parser.cs:line 54
   at ParserTests.EarlyEos.EarlyEosTests.Parse(String query) in csly/ParserTests/EarlyEos/EarlyEosTests.cs:line 40
   at ParserTests.EarlyEos.EarlyEosTests.Early_EOS_NRE() in csly/ParserTests/EarlyEos/EarlyEosTests.cs:line 13
```

Can anyone help me figure out why the lexer is getting that EOS?

I've created unit tests to reproduce the issue on my fork/branch: https://github.com/mauricio-morales/csly/tree/issue/223-early-eos

**Comment by b3b00** [comment]
I will look at it tokay. Thanks for providing a unit test.

**Comment by b3b00** [comment]
I found it,
First you've wanted to manage whitespaces in your grammar which is not an efficient way as it clutters your grammar with useless SPACE tokens (this is the general case when using parser generators). You'd better let CSLY regex lexer skip them during lexing phase.
To make CSLY regex lexer skip these tokens you have to mark your SPACE lexeme asskippable : 
```C#
[Lexeme(@"\s+",isSkippable:true)] // note the optional isSkippable parameter
        SPACE,
```
Then you can remove all references to SPACE in your grammar : 
```c#
using System;
using sly.lexer;
using sly.parser.generator;

namespace ParserTests.Issue223_EarlyEos
{
    public class EarlyEosParser
    {
        [Production("primary_expression: VALUE COLON primary_expression")]
        public FieldValueExpression PrimaryExpression(
            Token<EarlyEosToken> fieldNameToken,
            Token<EarlyEosToken> colonToken,
            Expression expression)
        {
            if (Enum.TryParse<Fields>(fieldNameToken.Value, true, out var fieldName))
            {
                return new FieldValueExpression(
                    fieldName,
                    expression
                );
            }

            return new FieldValueExpression(
                null,
                new ValueExpression($"{fieldNameToken.Value}:{expression}")
            );
        }

        [Production("primary_expression: LPAREN  expression  RPAREN")]
        public GroupExpression PrimaryExpression(
            Token<EarlyEosToken> lParenToken,
            Expression expression,            
            Token<EarlyEosToken> rParenToken
            )
        {
            return new GroupExpression(expression);
        }

        [Production("primary_expression: INT")]
        [Production("primary_expression: DOUBLE")]
        [Production("primary_expression: VALUE")]
        [Production("primary_expression: QUOTED_VALUE")]
        public ValueExpression PrimaryExpression(Token<EarlyEosToken> valueToken)
        {
            switch (valueToken.TokenID)
            {
                case EarlyEosToken.INT:
                case EarlyEosToken.DOUBLE:
                case EarlyEosToken.VALUE:
                    return new ValueExpression(valueToken.Value);
                case EarlyEosToken.QUOTED_VALUE:
                    return new ValueExpression(
                        valueToken.Value
                            .TrimStart('\'')
                            .TrimStart('"')
                            .TrimEnd('\'')
                            .TrimEnd('"')
                    );
                default:
                    throw new Exception($"Invalid value token {valueToken.TokenID}");
            }
        }

        [Production("secondary_expression: primary_expression")]
        public Expression SecondaryExpression(Expression valueExpression) => valueExpression;

        [Production("secondary_expression: NOT  secondary_expression")]
        [Production("secondary_expression: PLUS secondary_expression")]
        [Production("secondary_expression: MINUS secondary_expression")]
        public Expression SecondaryExpression(
            Token<EarlyEosToken> unaryOperatorToken,
            Expression unaryExpression)
        {
            var unaryOperator = Enum.Parse<UnaryOperator>(unaryOperatorToken.Value, true);
            return new UnaryExpression(unaryOperator, unaryExpression);
        }

        [Production("tertiary_expression: secondary_expression")]
        public Expression TertiaryExpression(Expression valueExpression) => valueExpression;

        [Production("tertiary_expression: secondary_expression  tertiary_expression")]
        public BinaryExpression TertiaryExpression(
            Expression leftExpression,
            Expression rightExpression)
        {
            return new BinaryExpression(leftExpression, BinaryOperator.AND, rightExpression);
        }

        [Production("tertiary_expression: secondary_expression  AND  tertiary_expression")]
        [Production("tertiary_expression: secondary_expression  OR  tertiary_expression")]
        public BinaryExpression BinaryExpression(
            Expression leftExpression,
            Token<EarlyEosToken> binaryOperatorToken,
            Expression rightExpression)
        {
            var binaryOperator = Enum.Parse<BinaryOperator>(binaryOperatorToken.Value, true);
            return new BinaryExpression(leftExpression, binaryOperator, rightExpression);
        }

        [Production("expression: tertiary_expression")]
        public Expression Expression(Expression valueExpression) => valueExpression;
    }}
``` 

Tell me if this answer is good for you

**Comment by b3b00** [comment]
As a side note I would advise you to 
 - use the (https://github.com/b3b00/csly/wiki/expression-parsing)[ expression parsing] feature to easen expression parsing and operation precedence.
 - use the [generic lexer](https://github.com/b3b00/csly/wiki/GenericLexer) instead of the [regex lexer](https://github.com/b3b00/csly/wiki/RegexLexer) for :  
     - better general performance
     - easier use : my personal taste

**Comment by b3b00** [comment]
I will keep you unit test as I think CSLY should not crash as it does in your case. Maybe I will come with a fix later.

**Comment by mauricio-morales** [comment]
> I found it,
> First you've wanted to manage whitespaces in your grammar which is not an efficient way as it clutters your grammar with useless SPACE tokens (this is the general case when using parser generators). You'd better let CSLY regex lexer skip them during lexing phase.
> To make CSLY regex lexer skip these tokens you have to mark your SPACE lexeme asskippable :
> 
> ```cs
> [Lexeme(@"\s+",isSkippable:true)] // note the optional isSkippable parameter
>         SPACE,
> ```
> 
> Then you can remove all references to SPACE in your grammar :
> 
> ```cs
> using System;
> using sly.lexer;
> using sly.parser.generator;
> 
> namespace ParserTests.Issue223_EarlyEos
> {
>     public class EarlyEosParser
>     {
>         [Production("primary_expression: VALUE COLON primary_expression")]
>         public FieldValueExpression PrimaryExpression(
>             Token<EarlyEosToken> fieldNameToken,
>             Token<EarlyEosToken> colonToken,
>             Expression expression)
>         {
>             if (Enum.TryParse<Fields>(fieldNameToken.Value, true, out var fieldName))
>             {
>                 return new FieldValueExpression(
>                     fieldName,
>                     expression
>                 );
>             }
> 
>             return new FieldValueExpression(
>                 null,
>                 new ValueExpression($"{fieldNameToken.Value}:{expression}")
>             );
>         }
> 
>         [Production("primary_expression: LPAREN  expression  RPAREN")]
>         public GroupExpression PrimaryExpression(
>             Token<EarlyEosToken> lParenToken,
>             Expression expression,            
>             Token<EarlyEosToken> rParenToken
>             )
>         {
>             return new GroupExpression(expression);
>         }
> 
>         [Production("primary_expression: INT")]
>         [Production("primary_expression: DOUBLE")]
>         [Production("primary_expression: VALUE")]
>         [Production("primary_expression: QUOTED_VALUE")]
>         public ValueExpression PrimaryExpression(Token<EarlyEosToken> valueToken)
>         {
>             switch (valueToken.TokenID)
>             {
>                 case EarlyEosToken.INT:
>                 case EarlyEosToken.DOUBLE:
>                 case EarlyEosToken.VALUE:
>                     return new ValueExpression(valueToken.Value);
>                 case EarlyEosToken.QUOTED_VALUE:
>                     return new ValueExpression(
>                         valueToken.Value
>                             .TrimStart('\'')
>                             .TrimStart('"')
>                             .TrimEnd('\'')
>                             .TrimEnd('"')
>                     );
>                 default:
>                     throw new Exception($"Invalid value token {valueToken.TokenID}");
>             }
>         }
> 
>         [Production("secondary_expression: primary_expression")]
>         public Expression SecondaryExpression(Expression valueExpression) => valueExpression;
> 
>         [Production("secondary_expression: NOT  secondary_expression")]
>         [Production("secondary_expression: PLUS secondary_expression")]
>         [Production("secondary_expression: MINUS secondary_expression")]
>         public Expression SecondaryExpression(
>             Token<EarlyEosToken> unaryOperatorToken,
>             Expression unaryExpression)
>         {
>             var unaryOperator = Enum.Parse<UnaryOperator>(unaryOperatorToken.Value, true);
>             return new UnaryExpression(unaryOperator, unaryExpression);
>         }
> 
>         [Production("tertiary_expression: secondary_expression")]
>         public Expression TertiaryExpression(Expression valueExpression) => valueExpression;
> 
>         [Production("tertiary_expression: secondary_expression  tertiary_expression")]
>         public BinaryExpression TertiaryExpression(
>             Expression leftExpression,
>             Expression rightExpression)
>         {
>             return new BinaryExpression(leftExpression, BinaryOperator.AND, rightExpression);
>         }
> 
>         [Production("tertiary_expression: secondary_expression  AND  tertiary_expression")]
>         [Production("tertiary_expression: secondary_expression  OR  tertiary_expression")]
>         public BinaryExpression BinaryExpression(
>             Expression leftExpression,
>             Token<EarlyEosToken> binaryOperatorToken,
>             Expression rightExpression)
>         {
>             var binaryOperator = Enum.Parse<BinaryOperator>(binaryOperatorToken.Value, true);
>             return new BinaryExpression(leftExpression, binaryOperator, rightExpression);
>         }
> 
>         [Production("expression: tertiary_expression")]
>         public Expression Expression(Expression valueExpression) => valueExpression;
>     }}
> ```
> 
> Tell me if this answer is good for you

This worked perfectly! thanks! I had forced the `SPACE` token because this query language has a construct `+term` and `-term` to include/exclude from the search... so I thought I needed the `SPACE` everywhere else to make it clear there was no space in those constructs... but with your suggestion all tests pass so looks like this fixed it!

I'll implement your other suggestions in a bit! Thank you so much! I love this library!

**Comment by b3b00** [comment]
@mauricio-morales Sorry I deleted your other issue. Still the fix is available on branche bugifix/223.
I 'm still investigating on this #223 issue.

**Comment by b3b00** [comment]
Hello @mauricio-morales , I've found and fixed the issue. You can try it with a nuget  [sly 2.6.4-beta](https://www.nuget.org/packages/sly/2.6.4-beta).
Could you try and tell me before  I merge to dev branch and release a new sly version ?

Thanks for having reporting the issue it is really strange that it did not appear before.

Olivier

**Comment by mauricio-morales** [comment]
Thanks @b3b00! I did test beta and the error from #225 is no longer there. I still have ways to go before I can get the GenericLexer working and will probably have to backlog that work cause I'm running behind, but I can confirm that your beta did get rid of the error without any change to my grammar. 

PS: While I have your attention, you may want to consider https://github.com/b3b00/csly/pull/227 before releasing a new sly version.

**Comment by b3b00** [comment]
I released v2.6.4 including fixes for your issues along with your regex lexer improvment PR. Tanks for both
